socketio/src/socket.rs

```diff
@@ -585,19 +533,44 @@ impl<'a> Iterator for Iter<'a> {
             let next = next.unwrap();
 
             match next.packet_id {
-                EnginePacketId::MessageBase64 => {
-                    // This is an attachment
-                    let packet = self.socket.handle_new_message(next.data);
-                    if let Some(packet) = packet {
-                        return Some(Ok(packet));
+                //TODO: refactor me
+                EnginePacketId::MessageBinary | EnginePacketId::Message => {
+                    let socket_packet = SocketPacket::try_from(&next.data);
+                    if let Err(err) = socket_packet {
+                        return Some(Err(err));
                     }
-                }
-                EnginePacketId::Message => {
-                    let packet = self.socket.handle_new_message(next.data);
+                    // SAFETY: checked above to see if it was Err
+                    let mut socket_packet = socket_packet.unwrap();
+                    // Only handle attachments if there are any
+                    if socket_packet.attachment_count > 0 {
+                        let mut attachments_left = socket_packet.attachment_count;
+                        let mut attachments = Vec::new();
+                        while attachments_left > 0 {
+                            let next = self.engine_iter.next()?;
+                            if let Err(err) = next {
+                                return Some(Err(err.into()));
+                            }
+                            // SAFETY: Checked to see if it was error above.
+                            let packet = next.unwrap();
+                            match packet.packet_id {
+                                EnginePacketId::MessageBinary | EnginePacketId::Message => {
+                                    attachments.push(packet.data);
+                                    attachments_left = attachments_left - 1;
+                                }
+                                _ => {
+                                    return Some(Err(Error::InvalidAttachmentPacketType(
+                                        packet.packet_id.into(),
+                                    )));
+                                }
+                            }
```

Comment:
This could be collided to a match block instead of unwrapping after the if let.

Reply:
Originally had it like that, but tab level was getting out of hand. a03ad2f6ed020cd96e56eb53bec7314923d8d049.
Refactored into a handle_new_entineio_packet function in 6c3b48c04d82f60c00ec039823902fa45b5aeeb2
Will be refactoring this section further in PRs to come, out of scope for this PR (added to checklist).

Reply:
Alright!